File: Leetcode/Array/011_Container_With_Most_Water.java
public class Solution {
    // O(n) time, O(1) space
    public int maxArea(int[] height) {
        if(height == null || height.length == 0)
            return 0;
        int l = 0, r = height.length - 1;
        int area = Integer.MIN_VALUE;
        while(l < r){
            int h = Math.min(height[l], height[r]);
            area = Math.max(area, (r - l) * h);
            if(height[l] < height[r])   l++;
            else r--;
        }
        return area;
    }
}

File: Leetcode/All/011_Container_With_Most_Water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int i = 0, j = height.size() - 1;
        int area = 0;
        while(i < j){
            int h = min(height[i], height[j]);
            area = max(area, h * (j - i));
            if(height[i] > height[j])
                j--;
            else
                i++;
        }
        return area;
    }
};